### PR TITLE
`write_cassandra` block #124

### PR DESCRIPTION
**Pull request by spicy-sauce**

### CODE REVIEW

File: core/src/datayoga_core/write_utils.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+from typing import Any, Callable, Dict, List, Union
+
+from datayoga_core.block import Block
+from datayoga_core.opcode import OpCode
+from datayoga_core.result import Result, Status
+
+logger = logging.getLogger("dy")
+
+
+def write_records(data: List[Dict[str, Any]],
```
Comment (zalmane) [inline core/src/datayoga_core/write_utils.py]: I would prefer not to send Callables as method parameters. It can get very messy very quickly.
Why not have this method just do the batching for you, and return 3 batches. The block can then decide what to do with it. We can then implement the deduplication here too.

File: core/src/datayoga_core/write_utils.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            execute_delete(records)
+        else:
+            for record in records:
+                record[Block.RESULT_FIELD] = Result(Status.REJECTED, f"{opcode} - unsupported opcode")
+            logger.warning(f"{opcode} - unsupported opcode")
+
+    return data
+
+
+def get_source_column(item: Union[Dict[str, str], str]) -> str:
```
Comment (zalmane) [inline core/src/datayoga_core/write_utils.py]: what do these methods do? Do we really need them as a utility method? 
in any case, I would have clear comments since it is hard to follow what is their purpose

File: core/src/datayoga_core/write_utils.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+def get_target_column(item: Union[Dict[str, str], str]) -> str:
+    return next(iter(item.keys())) if isinstance(item, dict) else item
+
+
+def get_column_mapping(mapping: List[Union[Dict[str, str], str]]) -> List[Dict[str, str]]:
+    return [{"column": next(iter(item.keys())), "key": next(iter(item.values()))} if isinstance(item, dict)
+            else {"column": item, "key": item} for item in mapping] if mapping else []
+
+
+def get_key_values(record: Dict[str, Any], keys: List[Union[Dict[str, Any], str]]) -> Dict[str, Any]:
```
Comment (zalmane) [inline core/src/datayoga_core/write_utils.py]: please add clear comment explaining the purpose

File: core/src/datayoga_core/write_utils.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            source = get_source_column(item)
+            target = get_target_column(item)
+            record_to_upsert[target] = None if source not in record else record[source]
+
+        records_to_upsert.append(record_to_upsert)
+
+    return records_to_upsert
+
+
+def get_records_to_delete(records: List[Dict[str, Any]],
```
Comment (zalmane) [inline core/src/datayoga_core/write_utils.py]: aren't these a duplication of the above method that relates to upsert?

File: core/src/datayoga_core/blocks/cassandra/write/block.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+        pk_clause = f"{' and '.join([column + ' = ?' for column in business_key_columns])}"
+        self.delete_stmt = f"delete from {self.keyspace}.{self.table} where {pk_clause}"
+
+        self.upsert_stmt = f"update {self.keyspace}.{self.table} set {', '.join([column + ' = ?' for column in mapping_columns])} where {pk_clause}"
+
+    async def run(self, data: List[Dict[str, Any]]) -> Tuple[List[Dict[str, Any]], List[Result]]:
+        logger.debug(f"Running {self.get_block_name()}")
+
+        data = write_utils.write_records(data, self.opcode_field, self.execute_upsert, self.execute_delete)
```
Comment (zalmane) [inline core/src/datayoga_core/blocks/cassandra/write/block.py]: I assume we need to open a 'transaction' here? is there such a concept in cassandra?

File: core/src/datayoga_core/blocks/cassandra/write/block.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        return utils.produce_data_and_results(data)
+
+    def execute_upsert(self, records: List[Dict[str, Any]]):
+        records_to_upsert = write_utils.get_records_to_upsert(records, self.keys, self.mapping)
+
+        logger.debug(f"Upserting {len(records_to_upsert)} record(s) to {self.table} table")
+        if records_to_upsert:
+            batch = BatchStatement()
```
Comment (zalmane) [inline core/src/datayoga_core/blocks/cassandra/write/block.py]: I have some memory saying that if the keys are distributed, writing in batches may actually slow down performance:
https://stackoverflow.com/questions/65269666/how-do-i-configure-and-execute-batchstatement-in-cassandra-correctly

File: core/src/datayoga_core/write_utils.py (hunk 2 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/write_utils.py]: they take an object or a string of column mapping, for example:
{ "last_name": "lname" }
or 
"last_name"

And return `last_name` for source and `lname` for target.

Since we use them in a several places in `cassandra.write` and `relational.write`, I thought it would be good to have them reusable.

File: core/src/datayoga_core/write_utils.py (hunk 4 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/write_utils.py]: similar, but not the same, I guess I can unify them.

File: core/src/datayoga_core/blocks/cassandra/write/block.py (hunk 5 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/blocks/cassandra/write/block.py]: not the same thing, see:
https://docs.datastax.com/en/cassandra-oss/2.2/cassandra/dml/dmlTransactionsDiffer.html

File: core/src/datayoga_core/write_utils.py (hunk 4 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/write_utils.py]: done, now I have one named `map_records`

File: core/src/datayoga_core/write_utils.py (hunk 3 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/write_utils.py]: removed it completely.

File: core/src/datayoga_core/write_utils.py (hunk 2 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/write_utils.py]: removed them completely.

File: core/src/datayoga_core/write_utils.py (hunk 1 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/write_utils.py]: done

File: core/src/datayoga_core/blocks/cassandra/write/block.py (hunk 6 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/blocks/cassandra/write/block.py]: done, replaced with preparedstatement and execute_async

File: core/src/datayoga_core/blocks/cassandra/write/block.py [hunk 7]
```diff
@@ -54,27 +53,45 @@ def init(self, context: Optional[Context] = None):
...
+
+            logger.debug(f"Total {len(records)} record(s) with {opcode} opcode")
+            if opcode == OpCode.UPDATE.value:
+                self.execute_upsert(records)
+            elif opcode == OpCode.DELETE.value:
+                self.execute_delete(records)
+
         return utils.produce_data_and_results(data)
 
+    def get_future(self, stmt: PreparedStatement, record: Dict[str, Any]) -> Any:
```
Comment (zalmane) [inline core/src/datayoga_core/blocks/cassandra/write/block.py:61]: do we need a method for 2 lines if this is only called in one place? may be more maintainable to just inline

File: core/src/datayoga_core/write_utils.py [hunk 8]
```diff
@@ -1,92 +1,52 @@
...
-from datayoga_core.result import Result, Status
 
 logger = logging.getLogger("dy")
 
 
-def write_records(data: List[Dict[str, Any]],
-                  opcode_field: str, execute_upsert: Callable, execute_delete: Callable) -> List[Dict[str, Any]]:
-    # group records by opcode
-    opcodes = [{"opcode": key, "records": list(result)} for key, result in groupby(
+def group_records_by_opcode(data: List[Dict[str, Any]],
```
Comment (zalmane) [inline core/src/datayoga_core/write_utils.py:11]: can we just return 3 return variables instead of a list? is it always just three? In that case, would be easier than a return type of list

File: core/src/datayoga_core/blocks/cassandra/write/block.py (hunk 7 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/blocks/cassandra/write/block.py:61]: It's called from two places.

File: core/src/datayoga_core/write_utils.py (hunk 8 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/write_utils.py:11]: done

**Review by zalmane — APPROVED** [review]
